feat(experiments): validate events for Funnel experiments

jurajmajerik opened the pull request:
## Problem
Based on feedback from https://github.com/PostHog/posthog/pull/19969, splitting this implementation into 3 parts:
- Validate events for Funnel experiments (this PR)
- Validate events for Trends experiments (TODO)
- Style the validation error in the UI (TODO)

## Changes
- Check if the relevant events has been received directly in the `/experiments` endpoint. Return a `ValidationError` to be rendered in the UI:
<img width="804" alt="image" src="https://github.com/PostHog/posthog/assets/22996112/9bb82aca-63b0-4225-88f7-18737f503fe0">

- I left in place the existing validation in `calculate_results` as I can't rule out some edge cases will reach it. Happy to hear feedback.

## How did you test this code?
unit test

jurajmajerik committed:
add validation

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

neilkakkar commented on ee/clickhouse/queries/test/test_experiments.py:
```diff
@@ -0,0 +1,62 @@
+        self.assertIn(expected_message, str(context.exception))
+
+    def test_validate_event_variants_2(self):
+        filtered_results = [
```
Can we test this using real possible results from a funnel query? 🤔  Not sure such a result is ever possible

jurajmajerik replied:
Correct, this isn't possible. If step 1 is not present at all, the funnel results will be empty 👍

neilkakkar commented on ee/clickhouse/queries/experiments/funnel_experiment_result.py:
```diff
@@ -292,3 +296,27 @@ def calculate_probability_of_winning_for_each(variants: List[Variant]) -> List[P
+
+def validate_event_variants(filtered_results, variants):
+    if not filtered_results or not filtered_results[0]:
+        raise ValidationError("No experiment events have been ingested yet")
```
I'd _love_ to distinguish between the case where events are coming in without flag information, and where no events have been ingested, because they lead to very different user messages:

(1) is almost always implementation error / some action that the user needs to take (like updating code, or updating experiment metric)
(2) is almost always 'wait longer'

I think if you validate before filtering, we'll get this distinction.

neilkakkar commented on ee/clickhouse/queries/experiments/funnel_experiment_result.py:
```diff
@@ -292,3 +296,27 @@ def calculate_probability_of_winning_for_each(variants: List[Variant]) -> List[P
+    if not filtered_results or not filtered_results[0]:
+        raise ValidationError("No experiment events have been ingested yet")
+
+    eventsWithOrderZero = []
```
There needs to always be only one event with order 0

neilkakkar replied:
Sorry if this comment wasn't clear, I meant we don't need an array here, this should be just one specific event, keeps the data model clearer

neilkakkar replied:
Nah I'm confused, we'll have multiple when we breakdown, one for each breakdown value, so this makes sense, is correct 🤦

neilkakkar commented on ee/clickhouse/queries/experiments/funnel_experiment_result.py:
```diff
@@ -292,3 +296,27 @@ def calculate_probability_of_winning_for_each(variants: List[Variant]) -> List[P
+
+    if not len(missing_variants) == 0:
+        missing_variants_str = ", ".join(missing_variants)
+        message = f"No experiment events have been ingested yet for the following variants: {missing_variants_str}"
```
very nice :)

neilkakkar reviewed (commented):
Getting there! :)

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into feat/experiments-funnel-events-validation

jurajmajerik committed:
tweak validation

jurajmajerik committed:
Merge branch 'master' of https://github.com/PostHog/posthog into feat/experiments-funnel-events-validation

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots

jurajmajerik commented:
changes:
- moved the validation before filtering, so now we're validating "raw" funnel results
- checked that having `step 2` in funnel results without `step 1` is indeed not possible, removed that code path and test
- added error codes to the ValidationError's
- using real funnel query results in the tests

neilkakkar commented on ee/clickhouse/queries/experiments/funnel_experiment_result.py:
```diff
@@ -59,6 +59,7 @@ def __init__(
     ):
         breakdown_key = f"$feature/{feature_flag.key}"
         self.variants = [variant["key"] for variant in feature_flag.variants]
+        self.filter = filter
```
unnecessary?

neilkakkar commented on ee/clickhouse/queries/test/test_experiments.py:
```diff
@@ -0,0 +1,50 @@
+                    "type": "events",
+                    "average_conversion_time": None,
+                    "median_conversion_time": None,
+                    "breakdown": ["test"],
```
Can we have one more test please where the experiment cares about `test` and `control`, and the breakdown is `true`, and `bazinga` (random old variant) ?

neilkakkar commented on frontend/src/scenes/experiments/ExperimentResult.tsx:
```diff
@@ -183,9 +183,13 @@ export function ExperimentResult(): JSX.Element {
+                                        <b>There are no results for this experiment yet.</b>
+                                    </div>
+                                    {!!experimentResultCalculationError && (
+                                        <div className="text-sm mb-2">{experimentResultCalculationError}</div>
```
Assuming you'll revisit this after backend is done?

jurajmajerik replied:
yes indeed, what we have here is the current state of things (just render the error message in the results)

neilkakkar reviewed (approved):
Much nicer :)

Few comments, but ok to handle in a follow up!

jurajmajerik committed:
tweaks

github-actions[bot] committed:
Update query snapshots

github-actions[bot] committed:
Update query snapshots
